### PR TITLE
*: faq for snapshot warning and dynamically determining snapshotWarningTimeout

### PR DESCRIPTION
**Pull request by fanminshi**

### CODE REVIEW

File: mvcc/backend/backend.go [hunk 1]
```diff
@@ -119,6 +120,13 @@ func NewDefaultBackend(path string) Backend {
 }
 
 func newBackend(bcfg BackendConfig) *backend {
+	// snapshotWarningTimeout is based on transferring 4 GB over a 1 gigabit/s connection
+	// with max tcp throughput of 949.285 Mbit/s limited by TCP overhead.
+	// every 4 gigabytes increases snapshotWarningTimeout by 35 seconds.
+	// e.g size(db) <= 4 GB takes < 35 seconds, size(db) <= 8 GB takes < 70 seconds.
+	factor := bcfg.MmapSize/(4*1024*1024*1024) + 1
```
Comment (xiang90) [inline mvcc/backend/backend.go]: we can get the real db size at line 197. call tx.Size()

Comment (fanminshi) [inline mvcc/backend/backend.go]: @xiang90 is the idea giving warning base on transferring large db  multiple of 4 gigs, e.g 4 GB, 8 GB ...) file? If real db size is 10 mb, and it takes more than .5 second to transfer, do we care about giving a warning in this case?

Comment (xiang90) [inline mvcc/backend/backend.go]: probably have a min warning duration: 30seconds. if we know the db size is large, we increase 30 seconds to something more reasonable.

Comment (fanminshi) [inline mvcc/backend/backend.go]: @xiang90 I am thinking about calculating the warning duration base on following.

1. When Snapshot(), get real db size via `tx.Size()`.
2. `snapshotWarningTimeout` equals to either min warning duration or expected transfer time calculated using `tx.Size()`.

note: `snapshotWarningTimeout` becomes a local var of `Snapshot()`

Comment (xiang90) [inline mvcc/backend/backend.go]: sure.

**Comment by fanminshi** [comment]
All fixed. PTAL

File: mvcc/backend/backend.go [hunk 2]
```diff
@@ -165,31 +166,40 @@ func (b *backend) ForceCommit() {
...
+		plog.Fatalf("cannot begin tx (%s)", err)
+	}
+
 	stopc, donec := make(chan struct{}), make(chan struct{})
+	dbBytes := tx.Size()
 	go func() {
 		defer close(donec)
+		// snapshotWarningTimeout is based on transferring snapshot data over a 1 gigabit/s connection
+		// with max tcp throughput of 949.285 Mbit/s limited by TCP overhead.
+		var throughputBytes int64 = 118660625 // 949.285 Mbit/s = 118660625 bytes/s
```
Comment (xiang90) [inline mvcc/backend/backend.go]: i do not feel we need this accurate.

just do 100 * 1024 * 1024 // 100MB to be more conservative.

File: mvcc/backend/backend.go [hunk 3]
```diff
@@ -165,31 +166,40 @@ func (b *backend) ForceCommit() {
...
+			snapshotWarningTimeout = minSnapshotWarningTimeout
+		}
 		start := time.Now()
 		ticker := time.NewTicker(snapshotWarningTimeout)
 		defer ticker.Stop()
 		for {
 			select {
 			case <-ticker.C:
-				plog.Warningf("snapshotting is taking more than %v seconds to finish [started at %v]", time.Since(start).Seconds(), start)
+				plog.Warningf("snapshotting is taking more than %v seconds to finish transferring %v bytes [started at %v]", time.Since(start).Seconds(), dbBytes, start)
```
Comment (xiang90) [inline mvcc/backend/backend.go]: convert bytes into MB to be more human-readable?

**Comment by xiang90** [comment]
defer to @gyuho and @heyitsanthony

File: Documentation/faq.md [hunk 4]
```diff
@@ -118,6 +118,12 @@ Every new etcd cluster generates a new cluster ID based on the initial cluster c
 
 Usually this warning happens after tearing down an old cluster, then reusing some of the peer addresses for the new cluster. If any etcd process from the old cluster is still running it will try to contact the new cluster. The new cluster will recognize a cluster ID mismatch, then ignore the request and emit this warning. This warning is often cleared by ensuring peer addresses among distinct clusters are disjoint.
 
+#### What does the etcd warning "snapshotting is taking more than x seconds to finish ..." mean?
+
+etcd can transfer its key value store as a snapshot for [backup][backup] or to a slow follower for bring it up to speed. The expected transfer time is based on transferring snapshot over a 1 gigabit/s connection assuming a minimum tcp throughput of 100 MB/s. When the actual transfer time is greater than min(30 seconds, expected transfer time), the warning appears. The warning allows one to reason about the etcd snapshotting performance. If one expects a slow snapshotting performance due to a slow underlying infrastructure, then ignore this warning. If the performance is slower than expected, use this warning for debugging/investigating.
```
Comment (heyitsanthony) [inline Documentation/faq.md]: etcd sends a snapshot of its complete key-value store to refresh slow followers and for [backups][backup]. Slow snapshot transfer times increase MTTR; if the cluster is ingesting data with high throughput, slow followers may livelock by needing a new snapshot before finishing receiving a snapshot. To catch slow snapshot performance, etcd warns when sending a snapshot takes more than thirty seconds and exceeds the expected transfer time for a 1Gbps connection.

File: Documentation/faq.md [hunk 5]
```diff
@@ -118,6 +118,12 @@ Every new etcd cluster generates a new cluster ID based on the initial cluster c
 
 Usually this warning happens after tearing down an old cluster, then reusing some of the peer addresses for the new cluster. If any etcd process from the old cluster is still running it will try to contact the new cluster. The new cluster will recognize a cluster ID mismatch, then ignore the request and emit this warning. This warning is often cleared by ensuring peer addresses among distinct clusters are disjoint.
 
+#### What does the etcd warning "snapshotting is taking more than x seconds to finish ..." mean?
+
+etcd can transfer its key value store as a snapshot for [backup][backup] or to a slow follower for bring it up to speed. The expected transfer time is based on transferring snapshot over a 1 gigabit/s connection assuming a minimum tcp throughput of 100 MB/s. When the actual transfer time is greater than min(30 seconds, expected transfer time), the warning appears. The warning allows one to reason about the etcd snapshotting performance. If one expects a slow snapshotting performance due to a slow underlying infrastructure, then ignore this warning. If the performance is slower than expected, use this warning for debugging/investigating. 
+
+
```
Comment (heyitsanthony) [inline Documentation/faq.md]: only one blank line please

File: mvcc/backend/backend.go [hunk 6]
```diff
@@ -165,31 +166,40 @@ func (b *backend) ForceCommit() {
...
+		plog.Fatalf("cannot begin tx (%s)", err)
+	}
+
 	stopc, donec := make(chan struct{}), make(chan struct{})
+	dbBytes := tx.Size()
 	go func() {
 		defer close(donec)
+		// snapshotWarningTimeout is based on transferring snapshot data over a 1 gigabit/s connection
+		// assuming a min tcp throughput of 100MB/s.
+		var throughputBytes int64 = 100 * 1024 * 1014
```
Comment (heyitsanthony) [inline mvcc/backend/backend.go]: if the constants change then the getting a 0 from integer division could break this code; use floats to be safe
```go
// assume 100MB/s transfer rates
sendRateBytes := int64(100 * 1024 * 1024)
warningTimeout := time.Duration((float64(dbBytes)/float64(sendRateBytes)) * float64(time.Second))
```

Comment (fanminshi) [inline mvcc/backend/backend.go]: `time.Duration((float64(dbBytes)/float64(sendRateBytes))` causes some kinda of warning;

`tmp/sandbox372115104/main.go:9: constant 0.1 truncated to integer`
https://play.golang.org/p/bokx5TeCvl


I think `warningTimeout := time.Duration(dbBytes/(sendRateBytes + 1)) *time.Second)` is probably good enough.

Comment (heyitsanthony) [inline mvcc/backend/backend.go]:
```go
time.Duration(int64((float64(dbBytes)/float64(sendRateBytes)) * float64(time.Second))
```

Comment (fanminshi) [inline mvcc/backend/backend.go]: divides by 0 error still occurs.
```go
d := time.Duration(int64((float64(1)/float64(0)) * float64(time.Second)))
fmt.Println(d)
```
https://play.golang.org/p/4FZHjQYKfh

Comment (heyitsanthony) [inline mvcc/backend/backend.go]: Huh? The concern is int/int will truncate to 0 for cases where float/float won't truncate to 0. Not concerned about sendRateBytes being 0 and therefore causing divide by zero.

Comment (fanminshi) [inline mvcc/backend/backend.go]: @heyitsanthony I see what you meant. Is there a issue with `warningTimeout` = 0? We have a `minSnapshotWarningTimeout` check.

Comment (heyitsanthony) [inline mvcc/backend/backend.go]: the expected rate will be 105MB/s when dbBytes % sendRate == sendRate-1

Comment (fanminshi) [inline mvcc/backend/backend.go]: got it.

**Comment by fanminshi** [comment]
All fixed. PTAL

**Comment by fanminshi** [comment]
unrelated Ci failure.

**Pull request merged**